src/mlpack/methods/ann/layer/bicubic_interpolation_impl.hpp

```diff
@@ -0,0 +1,229 @@
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BBICUBIC_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BICUBIC_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bicubic_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+BicubicInterpolation<InputDataType, OutputDataType>::
+BicubicInterpolation():
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  alpha(-0.75),
+  batchSize(0)
+{
+ // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BicubicInterpolation<InputDataType, OutputDataType>::
+BicubicInterpolation(const size_t inRowSize,
+                     const size_t inColSize,
+                     const size_t outRowSize,
+                     const size_t outColSize,
+                     const size_t depth,
+                     const double alpha):
+  inRowSize(inRowSize),
+  inColSize(inColSize),
+  outRowSize(outRowSize),
+  outColSize(outColSize),
+  depth(depth),
+  alpha(alpha),
+  batchSize(0)
+{
+
+}
+
+template<typename InputDataType, typename OutputDataType>
+void BicubicInterpolation<InputDataType, OutputDataType>::GetKernalWeight(double delta,
+                                                                          arma::mat& coeffs)
+{
+  coeffs(0) = ((alpha * (delta + 1) - 5 * alpha) * (delta + 1) + 8 * alpha) * (delta + 1) - 4 * alpha;
+  coeffs(1) = ((alpha + 2) * delta - (alpha + 3)) * delta * delta + 1;
+  coeffs(2) = ((alpha + 2) * (1 - delta) - (alpha + 3)) * (1 - delta) * (1 - delta) + 1;
+  coeffs(3) = 1 - coeffs[0] - coeffs[1] - coeffs[2];
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BicubicInterpolation<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  batchSize = input.n_cols;
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, batchSize);
+  else
+  {
+    assert(output.n_rows == outRowSize * outColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  const double scaleRow = (double) inRowSize / (double) outRowSize;
+  const double scaleCol = (double) inColSize / (double) outColSize;
+
+  arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+      inRowSize, inColSize, depth * batchSize, false, false);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
+      depth * batchSize, false, true);
+
+  for (size_t k = 0; k < depth * batchSize; ++k)
+  {
+    // The input is padded on all sides using replication of the input boundary.
+    arma::mat grid = arma::mat(inRowSize + 4, inColSize + 4);
+    grid.zeros();
```

Comment:
```suggestion
    arma::mat grid = arma::zeros(inRowSize + 4, inColSize + 4);
```

We can skip the extra step.